sdk/src/azure/core/az_mqtt5_connection_hfsm.c

```diff
@@ -311,30 +433,62 @@ static az_result reconnect_timeout(az_event_policy* me, az_event event)
   switch (event.type)
   {
     case AZ_HFSM_EVENT_ENTRY:
-      // TODO_L: reconnect timer (apply reconnect policy), counter not implemented.
+    {
+      _start_reconnect_timer(this_policy);
       break;
+    }
 
     case AZ_HFSM_EVENT_EXIT:
-      // TODO_L
+    {
+      _az_event_pipeline_timer* timer
+          = (_az_event_pipeline_timer*)&this_policy->_internal.connection_timer;
+      _az_FAULTED_IF_FAILED(
+          az_platform_timer_destroy(&timer->platform_timer), this_policy, reconnect_timeout);
       break;
+    }
 
     case AZ_HFSM_EVENT_TIMEOUT:
-      // TODO_L
+      if (event.data == &this_policy->_internal.connection_timer)
+      {
+        _az_FAULTED_IF_FAILED(
+            _az_hfsm_transition_peer((_az_hfsm*)me, reconnect_timeout, connecting),
+            this_policy,
+            reconnect_timeout);
+        _start_connect(this_policy);
+      }
+      break;
+
+    case AZ_EVENT_MQTT5_CONNECTION_CLOSE_REQ:
+      _az_FAULTED_IF_FAILED(
+          _az_hfsm_transition_superstate((_az_hfsm*)me, reconnect_timeout, started),
+          this_policy,
+          reconnect_timeout);
+      if (az_result_failed(_az_mqtt5_connection_api_callback(
+              this_policy, (az_event){ .type = AZ_MQTT5_EVENT_DISCONNECT_RSP, .data = NULL })))
+      {
+        // Callback failed: fault the connection object.
+        _az_RETURN_IF_FAILED(_az_hfsm_send_event(
+            (_az_hfsm*)me, (az_event){ .type = AZ_HFSM_EVENT_ERROR, .data = NULL }));
+      }
+      _az_RETURN_IF_FAILED(az_event_policy_send_inbound_event(
```

Comment:
I wonder if there's a way to have this get forwarded to the application from the other policies rather than sending it once to the application and once inbound

Reply:
We will have that in the future with a policy on the inbound end of the pipeline